packages/expo-av/ios/EXAV/AudioSampleCallback/EXAV+AudioSampleCallback.mm

```diff
@@ -0,0 +1,100 @@
+// Copyright 2021-present 650 Industries, Inc. (aka Expo)
+
+#import <Foundation/Foundation.h>
+
+#import <jsi/jsi.h>
+#import <ReactCommon/RCTTurboModuleManager.h>
+
+#import <EXAV/EXAV.h>
+#import <EXAV/EXAVPlayerData.h>
+#import <EXAV/EXAudioSampleCallback.h>
+#import <EXAV/AudioSampleCallbackWrapper.h>
+
+namespace jsi = facebook::jsi;
+using CallInvoker = facebook::react::CallInvoker;
+using namespace expo::av;
+
+static constexpr auto globalJsFuncName = "__EXAV_setOnAudioSampleReceivedCallback";
+static constexpr auto globalDestroyHostObjectName = "__EXAV_onDestroyHostObject";
+
+class CleanupOnDestroyHostObject : public jsi::HostObject {
+ public:
+  CleanupOnDestroyHostObject() {}
+  virtual ~CleanupOnDestroyHostObject() {
+    AudioSampleCallbackWrapper::removeAllCallbacks();
+  }
+  virtual jsi::Value get(jsi::Runtime &, const jsi::PropNameID &name) {
+    return jsi::Value::undefined();
+  }
+  virtual void set(jsi::Runtime &, const jsi::PropNameID &name, const jsi::Value &value) {}
+  virtual std::vector<jsi::PropNameID> getPropertyNames(jsi::Runtime &rt) {
+    return {};
+  }
+};
+
+@implementation EXAV (AudioSampleCallback)
+
+- (void)installJSIBindingsForRuntime:(void *)jsRuntimePtr
+                 withSoundDictionary:(NSMutableDictionary <NSNumber *, EXAVPlayerData *> *)soundDictionary
+{
+  __weak auto weakCallInvoker = self.bridge.jsCallInvoker;
+  __weak auto weakSoundDictionary = soundDictionary;
```

Comment:
Is it possible to create `EX_WEAKIFY`/`EX_STRONGIFY`/`EX_ENSURE_STRONGIFY_WITH_ERROR` macros for c++? 🤔

Reply:
I think it's not worth it:

Taking a look at the `EX_WEAKIFY` macro:
```objc
#define EX_WEAKIFY(var) __weak __typeof(var) EXWeak_##var = var;
```
I'd have to add 2 additional lines just to make them weak (instead of one `__weak` keyword):
```cpp
auto weakCallInvoker = self.bridge.jsCallInvoker;
auto weakSoundDictionary = soundDictionary;
EX_WEAKIFY(weakCallInvoker);
EX_WEAKIFY(weakSoundDictionary);
```

---

Also for `EX_ENSURE_STRONGIFY`, that would not let me throw the `jsi::JSError` if they couldn't be strongified, because of the macro:
```objc
#define EX_STRONGIFY(var) __strong __typeof(var) var = EXWeak_##var; 

#define EX_ENSURE_STRONGIFY(var) \
EX_STRONGIFY(var); \
if (var == nil) { return; }
```